### PR TITLE
Sign and publish Filecoin using hardware wallets

### PR DESCRIPTION
**Pull request by spylogsster**
Resolves https://github.com/brave/brave-browser/issues/22599

- Added signing and publishin Filecoin transactions with hardware wallets
  - Renamed `MethodNum` to `Method` for compatibility with `filecoin-wallet-provider` library
- Added support for UI network selector to import correct accounts for selected TEST/MAIN networks

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:
- Sign and publish FIL with Ledger device

### CODE REVIEW

**Comment by yrliou** [comment]
BTW, I only reviewed the backend part, please ask wallet frontend devs for the frontend part, thanks.

File: components/brave_wallet_ui/common/hardware/types.ts
Comment (petemill) [inline components/brave_wallet_ui/common/hardware/types.ts:72]: nit: is this name descriptive enough - I don't really understand what it is

**Review by petemill — APPROVED** [review]
JS deps update seems ok, but please consider whether a security review is needed for it. I'm not sure if they gave the ok for all updates to this repo or not.

File: components/brave_wallet/common/brave_wallet.mojom
Comment (spylogsster) [reply components/brave_wallet/common/brave_wallet.mojom:914]: > JS deps update seems ok, but please consider whether a security review is needed for it. I'm not sure if they gave the ok for all updates to this repo or not.

created, thanks https://github.com/brave/security/issues/864

**Review by Douglashdaniel — APPROVED** [review]
++

File: package.json
Comment (diracdeltas) [inline package.json:345]: any reason not to pin to a specific version here?

Comment (spylogsster) [reply package.json]: pinned, thanks!

File: components/brave_wallet/common/brave_wallet.mojom
Comment (spylogsster) [reply components/brave_wallet/common/brave_wallet.mojom:945]: needed to be able to pass callback to send transaction

File: components/brave_wallet/browser/fil_tx_manager.cc
Comment (spylogsster) [reply components/brave_wallet/browser/fil_tx_manager.cc]: done

File: components/brave_wallet/common/brave_wallet.mojom
Comment (spylogsster) [reply components/brave_wallet/common/brave_wallet.mojom]: done

File: components/brave_wallet_ui/common/hardware/types.ts
Comment (spylogsster) [reply components/brave_wallet_ui/common/hardware/types.ts]: done

File: components/brave_wallet_ui/common/hardware/ledgerjs/filecoin_ledger_keyring.ts
Comment (spylogsster) [reply components/brave_wallet_ui/common/hardware/ledgerjs/filecoin_ledger_keyring.ts]: done

File: components/brave_wallet_ui/common/hardware/ledgerjs/filecoin_ledger_keyring.ts
Comment (spylogsster) [reply components/brave_wallet_ui/common/hardware/ledgerjs/filecoin_ledger_keyring.ts]: done

**Review by yrliou — APPROVED** [review]
++ for C++ wallet backend part.

File: components/brave_wallet_ui/common/hardware/ledgerjs/filecoin_ledger_keyring.ts
Comment (kdenhartog) [inline components/brave_wallet_ui/common/hardware/ledgerjs/filecoin_ledger_keyring.ts:71]: Seeing this isn't changed so asking just for context. What motivated this to be the minLedgerVersion being set?

**Review by kdenhartog — COMMENTED** [review]
Overall this looks good from a security perspective. We should hold off on merging until @pes10k or @ShivanKaul get a chance to review the WebHID/WebUSB usage though.

File: package.json
Comment (spylogsster) [reply package.json]: done

File: components/brave_wallet_ui/common/hardware/ledgerjs/filecoin_ledger_keyring.ts
Comment (spylogsster) [reply components/brave_wallet_ui/common/hardware/ledgerjs/filecoin_ledger_keyring.ts:71]: it is required to be set by LedgerProvider

Comment (kdenhartog) [reply components/brave_wallet_ui/common/hardware/ledgerjs/filecoin_ledger_keyring.ts:71]: Ok, but we don't have a minimum then so just set it to the lowest possible value?

Comment (spylogsster) [reply components/brave_wallet_ui/common/hardware/ledgerjs/filecoin_ledger_keyring.ts:71]: yes

Comment (kdenhartog) [reply components/brave_wallet_ui/common/hardware/ledgerjs/filecoin_ledger_keyring.ts:71]: Awesome, thanks for the additional info

File: components/brave_wallet_ui/common/hardware/ledgerjs/filecoin_ledger_keyring.ts
Comment (kdenhartog) [reply components/brave_wallet_ui/common/hardware/ledgerjs/filecoin_ledger_keyring.ts:68]: Spoke with @pes10k and @fmarier about this and they said it's not something unexpected now. In the early days we disabled it, but have re-enabled support for them as the spec and APIs have improved and for webcompat reasons. Resolving this one with no action necessary

**Review by kdenhartog — APPROVED** [review]
All other suggested changes are optional nit's that I'll leave up to @spylogsster to choose if they want to implement them. This looks good from a sec perspective and my review of the dependency can be found in [this comment](https://github.com/brave/security/issues/864#issuecomment-1130254424)

File: components/brave_wallet_ui/common/async/hardware.ts
Comment (spylogsster) [reply components/brave_wallet_ui/common/async/hardware.ts:105]: done

File: components/brave_wallet_ui/common/async/hardware.ts
Comment (spylogsster) [reply components/brave_wallet_ui/common/async/hardware.ts:132]: done

File: components/brave_wallet_ui/common/hardware/ledgerjs/filecoin_ledger_keyring.ts
Comment (spylogsster) [reply components/brave_wallet_ui/common/hardware/ledgerjs/filecoin_ledger_keyring.ts]: done

File: components/brave_wallet/browser/fil_tx_manager.cc
Comment (yrliou) [inline components/brave_wallet/browser/fil_tx_manager.cc:290]: should be absl::nullopt instead

Comment (spylogsster) [reply components/brave_wallet/browser/fil_tx_manager.cc]: done

File: components/brave_wallet/browser/fil_transaction.cc
Comment (kdenhartog) [inline components/brave_wallet/browser/fil_transaction.cc:188]: Can you add a comment here saying something similar to what's documented on the `MPoolGetNonce` API? It says there, `if this is the first time wallet, this call will return an error. If the error contains "address not found", we assume the Nonce is 0, since no on-chain transactions reference this address.` 

I almost forgot that this is the way that API worked so thinking it would be useful for me if I come across this again later and forget.

Comment (spylogsster) [reply components/brave_wallet/browser/fil_transaction.cc:190]: done

**Comment by srirambv** [comment]
Verification passed on <html><head></head>
<body>


Brave | 1.41.11 Chromium: 102.0.5005.61 (Official Build) nightly (64-bit)
-- | --
Revision | `0e59bcc00cc4985ce39ad31c150065f159d95ad3-refs/branch-heads/5005@{#819}`
OS | Windows 11 Version 22H2 (Build 22621.1)


</body>
</html>

- Verified able to sign and publish Filecoin transactions using hardware wallets
- Verified transaction history is added to the submitted transaction
- Encountered [#23048](https://github.com/brave/brave-browser/issues/23048)

https://user-images.githubusercontent.com/17010094/170229803-ba1ebab5-c973-4b1f-9258-dc3dde28b31a.mov

**Pull request merged**
